File: week_3/보석 쇼핑.js
function solution(gems) {
	const types = new Set(gems);
	const n = gems.length;
	const k = types.size;

	const counter = [...types].reduce(
		(obj, key) => {
			obj[key] = 0;
			return obj;
		},
		{ __has: 0 }
	);
	let ep = -1;

	return gems.reduce(
		(answer, start, sp) => {
			while (ep < n - 1 && counter.__has != k) {
				const end = gems[++ep];
				if (!counter[end]++) counter.__has++;
			}

			if (ep - sp < answer[1] - answer[0] && counter.__has == k)
				answer = [sp + 1, ep + 1];

			if (!--counter[start]) counter.__has--;

			return answer;
		},
		[1, n]
	);
}

File: week_3/보석 쇼핑.py
from collections import defaultdict

def solution(gems):
    n, k = len(gems), len(set(gems))
    
    counter = defaultdict(int)
    subset = set()
    start = 0
    end = -1
    answer = [0, n]
    while start < n and end < n:
        if end - start < answer[1] - answer[0] and len(subset) == k:
            answer = [start + 1, end + 1]
        if end < n - 1 and len(subset) != k:
            end += 1
            counter[gems[end]] += 1
            subset.add(gems[end])
        else:
            counter[gems[start]] -= 1
            if not counter[gems[start]]:
                subset.discard(gems[start])
            start += 1
    
    return answer